### ISSUE TITLE
[Question] WhatsApp mode with paint editor

### ISSUE THREAD

**Issue by leandro-pilz** [hm21/pro_image_editor#58]
### Placeholder

Hello friend, a suggestion, I know that WhatsApp doesn't include circle, rectangular and other shapes, but it would be very interesting to have them, people loved the shapes, it would be very complicated, or there would be a way to customize this, I tried but it didn't work. I achieved

**Comment by hm21** [comment]
In the old versions, it was not possible to do this without getting offset problems from the layer position. However, I just released a new version for you `3.0.0-dev.20` as a prerelease, where you can add it like below. 

### Create your own widget
```dart
 configs: ProImageEditorConfigs(
      customWidgets: ImageEditorCustomWidgets(
        paintEditorBodyItem: Positioned(
          top: kToolbarHeight,
          left: 10,
          child: Theme(
            data: Theme.of(context),
            child: SingleChildScrollView(
              scrollDirection: Axis.vertical,
              child: Wrap(
                direction: Axis.vertical,
                spacing: 10,
                children: <Widget>[
                  ...List.generate(
                    paintModes.length,
                    (index) => Builder(
                      builder: (_) {
                        PaintModeBottomBarItem item = paintModes[index];
                        Color color = editorKey.currentState!.paintingEditor.currentState!.mode == item.mode
                            ? imageEditorPrimaryColor
                            : const Color(0xFFEEEEEE);

                        return Material(
                          color: Colors.transparent,
                          child: Ink(
                            decoration: const ShapeDecoration(
                              color: Colors.white30,
                              shape: CircleBorder(),
                            ),
                            child: IconButton(
                              color: Colors.white30,
                              icon: Icon(item.icon, color: color),
                              onPressed: () {
                                editorKey.currentState!.paintingEditor.currentState!.setMode(item.mode);
                              },
                            ),
                          ),
                        );
                      },
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ), 
),
```

### Add the paint modes you want
```dart
  List<PaintModeBottomBarItem> get paintModes => [
        const PaintModeBottomBarItem(
          mode: PaintModeE.freeStyle,
          icon: Icons.edit,
          label: 'Freestyle',
        ),
        const PaintModeBottomBarItem(
          mode: PaintModeE.arrow,
          icon: Icons.arrow_right_alt_outlined,
          label: 'Arrow',
        ),
        const PaintModeBottomBarItem(
          mode: PaintModeE.line,
          icon: Icons.horizontal_rule,
          label: 'Line',
        ),
        const PaintModeBottomBarItem(
          mode: PaintModeE.rect,
          icon: Icons.crop_free,
          label: 'Rectangle',
        ),
        const PaintModeBottomBarItem(
          mode: PaintModeE.circle,
          icon: Icons.lens_outlined,
          label: 'Circle',
        ),
        const PaintModeBottomBarItem(
          mode: PaintModeE.dashLine,
          icon: Icons.power_input,
          label: 'DashLine',
        ),
      ];
```

By the way, I created a new tab [discussions](https://github.com/hm21/pro_image_editor/discussions) where you can create new questions. I would love it if you could post your future questions there, that way it will be easier for other users to read those questions if they have the same problem :)

**Comment by leandro-pilz** [comment]
Hello, right friend, I will post it there, I will even ask one more
question there, as we really liked the idea of using this WhatsApp mode and
I would have one more suggestion to make.

Em ter., 28 de mai. de 2024 às 14:15, Alex Frei ***@***.***>
escreveu:

**Comment by leandro-pilz** [comment]
Hi, sorry to disturb you, I know it's a pre-release version, but when you
click on one of the editing options, I circle, rectangle... it doesn't
change the color of the icon to dark blue, just as an adjustment idea.
Thank you very much for your help, the support you provide is fantastic and
the package is really good, congratulations!!!

The image below with the side menu on the left now has the circle in blue,
but if I click on another color, it doesn't change color, if I use hot
reload it changes.

[image: image.png]

Em ter., 28 de mai. de 2024 às 14:15, Alex Frei ***@***.***>
escreveu:

**Comment by leandro-pilz** [comment]
Sorry friend for being in the way, but the problem with asuka is in version
3.0.0-dev.20

https://github.com/hm21/pro_image_editor/issues/48

Em ter., 28 de mai. de 2024 às 14:15, Alex Frei ***@***.***>
escreveu:

**Comment by hm21** [comment]
Ah, I see, there is a problem with the state management because the widget is in the three above the paint editor. You can fix it if you use a `StatefulBuilder` like below:

```dart
SingleChildScrollView(
    scrollDirection: Axis.vertical,
    child: StatefulBuilder(builder: (context, setState) {
      return Wrap(
        direction: Axis.vertical,
        spacing: 10,
        children: <Widget>[
          ...List.generate(
            paintModes.length,
            (index) => Builder(
              builder: (_) {
                PaintModeBottomBarItem item = paintModes[index];
                Color color = editorKey.currentState!.paintingEditor.currentState!.mode == item.mode
                    ? imageEditorPrimaryColor
                    : const Color(0xFFEEEEEE);

                return Material(
                  color: Colors.transparent,
                  child: Ink(
                    decoration: const ShapeDecoration(
                      color: Colors.white30,
                      shape: CircleBorder(),
                    ),
                    child: IconButton(
                      color: Colors.white30,
                      icon: Icon(item.icon, color: color),
                      onPressed: () {
                        editorKey.currentState!.paintingEditor.currentState!.setMode(item.mode);
                        editorKey.currentState!.paintingEditor.currentState!.setState(() {});
                        setState(() {});
                      },
                    ),
                  ),
                );
              },
            ),
          ),
        ],
      );
    }),
  ),
```

Alternatively, you can use a `Streambuilder` like in the example [here](https://github.com/hm21/pro_image_editor/blob/eb1bad4a18e4e43579566c0d5f437dc0a507fc72/example/lib/pages/custom_appbar_bottombar_example.dart) with a custom appbar/bottombar.


The issue with Asuka I will try, but actually the fix I made is also in the dev version. Does it make it again problem with the loading-dialog or in any other situation?

**Comment by hm21** [comment]
Okay, I found the problem with Asuka, but I will need time to resolve this issue. The issue is because in this prerelease the image editor is much faster than before. So now the editor generates the final image faster than it can open the loading-dialog. Before, I resolved this issue with using the context from the dialog itself, but now there is no context anymore (only if I slow down the editor). 
However, I will find another solution and give you an answer here then. 

Btw, when I checked the old issue of Asuka, I saw in your picture that you didn't add the `asukaHeroController` to the MaterialApp widget. I'm not sure if you want to disable hero animations, but if not, you can add this option `navigatorObservers: [Asuka.asukaHeroController]` to your MaterialApp widget that the hero animations will work with Asuka.

**Comment by leandro-pilz** [comment]
Perfect, about Asuka's hero, yes, I ended up adding it later, thanks for
the observation, well, I'm going to do the following for now, I saw that in
your example you can customize the app bar and bottom bar, I'm going to
create a customized situation here in parallel, and so on If you have
adjusted it, I will look for the solution in the package itself. Thank you
very much for all your help and attention.

Em ter., 28 de mai. de 2024 às 17:07, Alex Frei ***@***.***>
escreveu:

**Comment by hm21** [comment]
Hey Leandro,

Good news, I released version `3.0.0-dev.22` which should resolve the problem with Asuka. The issue was related to waiting for the loading dialog with its context. I optimized this as much as possible, but it will slow down the generation process by about 10ms - 20ms, which, I think, is acceptable. To enable this special generation process where it waits for the loading dialog context, you need to set `awaitLoadingDialogContext` to `true` in the configuration, as shown below:

```dart
configs: const ProImageEditorConfigs(
    imageGenerationConfigs: const ImageGeneratioConfigs(
      awaitLoadingDialogContext: true,
    ),
  ),
```

Btw, keep in mind that if you use the `LoadingDialog` outside the editor like I do in the examples, you also need to set `awaitLoadingDialogContext` to `true` in the configs there.

**Comment by leandro-pilz** [comment]
Perfect my friend, really good, well I ended up customizing it by looking
at the examples you have on github, and with this I'm delivering the app
feature, and it looks really cool too. But when you have the stable version
containing all this, possibly 3.0.0, I will do the tests to use the
solution natively. Thank you very much for your help, your work and the
package are very good.

I did the customization because as I still need to deliver the project
today, it was the best thing I could think of. Thanks a lot for the help.

Em qua., 29 de mai. de 2024 às 04:27, Alex Frei ***@***.***>
escreveu:
